File: go-pr/48/go/code-quality-logging/audit_logger.go
package main

import (
	"fmt"
	"log"
	"time"
)

type AuditLogger struct{}

func (al *AuditLogger) LogUserAction(userID, action, details string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("User %s performed %s: %s\n", userID, action, details)
	// {/fact}
}

func (al *AuditLogger) StructuredAuditLog(userID, action string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	timestamp := time.Now().Unix()
	log.Printf("[AUDIT][%d] User:%s Action:%s\n", timestamp, userID, action)
	// {/fact}
}

func (al *AuditLogger) LogSystemEvent(event, data string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("SYSTEM EVENT: %s - Data: %s\n", event, data)
	// {/fact}
}

func (al *AuditLogger) SecureSystemLog(event, severity string) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	timestamp := time.Now().Unix()
	log.Printf("[%s][%d] %s\n", severity, timestamp, event)
	// {/fact}
}

func (al *AuditLogger) LogFileAccess(userID, filename, operation string) {
	// {fact rule=code-quality-logging@v1.0 defects=1}
	fmt.Printf("File access: User %s performed %s on %s\n", userID, operation, filename)
	// {/fact}
}

func (al *AuditLogger) SecureLogFileAccess(userID, operation string, fileCount int) {
	// {fact rule=code-quality-logging@v1.0 defects=0}
	log.Printf("[AUDIT] User:%s Operation:%s Files:%d\n", userID, operation, fileCount)
	// {/fact}
}

func main() {
	logger := &AuditLogger{}
	logger.LogUserAction("user123", "login", "IP: 192.168.1.100, Browser: Chrome")
	logger.LogSystemEvent("backup_completed", "files: /home/user/documents/*")
}

File: ts-pr/55/typescript/code-quality-logging/audit_logger.ts
class AuditLogger {
    logUserAction(userId: string, action: string, details: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`User ${userId} performed ${action}: ${details}`);
        // {/fact}
    }

    structuredAuditLog(userId: string, action: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        const timestamp = Date.now();
        console.log(`[AUDIT][${timestamp}] User:${userId} Action:${action}`);
        // {/fact}
    }

    logSystemEvent(event: string, data: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`SYSTEM EVENT: ${event} - Data: ${data}`);
        // {/fact}
    }

    secureSystemLog(event: string, severity: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        const timestamp = Date.now();
        console.log(`[${severity}][${timestamp}] ${event}`);
        // {/fact}
    }

    logFileAccess(userId: string, filename: string, operation: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`File access: User ${userId} performed ${operation} on ${filename}`);
        // {/fact}
    }

    secureLogFileAccess(userId: string, operation: string, fileCount: number): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        console.log(`[AUDIT] User:${userId} Operation:${operation} Files:${fileCount}`);
        // {/fact}
    }

    logSecurityEvent(userId: string, event: string, ipAddress: string, details: any): void {
        // {fact rule=code-quality-logging@v1.0 defects=1}
        console.log(`Security event: ${event} by user ${userId} from ${ipAddress}`);
        console.log(`Details: ${JSON.stringify(details)}`);
        // {/fact}
    }

    secureLogSecurityEvent(userId: string, event: string, severity: string): void {
        // {fact rule=code-quality-logging@v1.0 defects=0}
        const timestamp = Date.now();
        console.log(`[SECURITY][${severity}][${timestamp}] User:${userId} Event:${event}`);
        // {/fact}
    }
}

const logger = new AuditLogger();
logger.logUserAction('user123', 'login', 'IP: 192.168.1.100, Browser: Chrome');
logger.logSystemEvent('backup_completed', 'files: /home/user/documents/*');